### PR TITLE
Add Support for MT5 model on Gaudi2

### PR DESCRIPTION
**Pull request by Gaurav7888**
Added Support for MT5 model on Gaudi2 using Static Shape Concept.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by vivekgoe** [comment]
@Gaurav7888 please check tests for translation using T5 here https://github.com/huggingface/optimum-habana/blob/main/tests/test_examples.py . Can you please add a test for MT5 also? Thanks.

**Review by ssarkar2 — CHANGES_REQUESTED** [review]
Please add tests and could you also mention the cmd line you used to test this in the PR

**Review by libinta — COMMENTED** [review]
@Gaurav7888  can you please address the comments?

**Comment by Gaurav7888** [comment]
@vivekgoe @ssarkar2 @libinta I have added the test files for mt5 model.

**Comment by ssarkar2** [comment]
@Gaurav7888 , there is a conflict that needs to be resolved. Please merge the latest main and your branch and resolve the conflicts

**Review by libinta — COMMENTED** [review]
@Gaurav7888  please provide the cmd test test in readme, also how is perf comparison with gpu?

**Comment by Gaurav7888** [comment]
@ssarkar2 I have made the changes and resolved the conflict. please review.

**Comment by libinta** [comment]
@Gaurav7888  can you resolve the conflicts with main branch?

**Comment by Gaurav7888** [comment]
> @Gaurav7888 can you resolve the conflicts with main branch?

Please review it, I have made changes.

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 1]
```diff
@@ -0,0 +1,640 @@
...
+        key_value_states=None,
+        position_bias=None,
+        past_key_value=None,
+        layer_head_mask=None,
+        query_length=None,
+        use_cache=False,
+        output_attentions=False,
+        token_idx=None,
+    ):
+        """
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:63]: @Gaurav7888 same as above

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 2]
```diff
@@ -0,0 +1,640 @@
+import warnings
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:1]: @Gaurav7888 you have missed the header, which includes the model name and copyright

**Comment by yafshar** [comment]
@Gaurav7888 please provide the test cmd in the README, also how is perf in comparison with gpu?

**Comment by yafshar** [comment]
@Gaurav7888  Please also run make style

File: tests/transformers/tests/models/mt5/test_modeling_t5.py [hunk 3]
```diff
@@ -0,0 +1,1047 @@
...
+        AutoTokenizer,
+        MT5EncoderModel,
+        MT5ForConditionalGeneration,
+        MT5ForQuestionAnswering,
+        MT5ForSequenceClassification,
+        MT5Model,
+    )
+
+from transformers.models.t5.modeling_t5 import T5_PRETRAINED_MODEL_ARCHIVE_LIST
+torch_device = "hpu"
```
Comment (yafshar) [inline tests/transformers/tests/models/mt5/test_modeling_t5.py:55]: @Gaurav7888 why are you overwriting the `torch_device` it is already imported in L28. If this is for your own testing please remove it

File: tests/transformers/tests/models/mt5/test_modeling_t5.py [hunk 4]
```diff
@@ -0,0 +1,1047 @@
...
+            attention_mask,
+            decoder_attention_mask,
+            lm_labels,
+        )
+
+    def test_decoder_model_past_with_large_inputs(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_decoder_model_past_large_inputs(*config_and_inputs)
+
+    def test_generate_with_past_key_values(self):
```
Comment (yafshar) [inline tests/transformers/tests/models/mt5/test_modeling_t5.py:806]: @Gaurav7888 this test is failing! Does it work correctly on your end. Please check

Comment (yafshar) [reply tests/transformers/tests/models/mt5/test_modeling_t5.py:806]: Please fix the file name `test_modeling_t5.py` -> `test_modeling_mt5.py`

**Comment by Gaurav7888** [comment]
@yafshar  Hi, I have made the changes as requested. Please review

**Comment by yafshar** [comment]
> @yafshar Hi, I have made the changes as requested. Please review

@Gaurav7888 thanks, please provide your test cmd?

**Comment by Gaurav7888** [comment]
@yafshar 

python run_translation.py \
    --model_name_or_path mt5-small \
    --do_eval \
    --source_lang en \
    --target_lang ro \
    --source_prefix "translate English to Romanian: " \
    --dataset_name wmt16 \
    --dataset_config_name ro-en \
    --output_dir /tmp/tst-translation \
    --per_device_eval_batch_size 4 \
    --overwrite_output_dir \
    --predict_with_generate \
    --use_habana \
    --use_lazy_mode \
    --use_hpu_graphs_for_inference \
    --gaudi_config_name Habana/t5 \
    --ignore_pad_token_for_loss False \
    --pad_to_max_length \
    --bf16

Is this you are asking? Please correct me.

File: tests/transformers/tests/models/mt5/test_modeling_t5.py
Comment (yafshar) [inline tests/transformers/tests/models/mt5/test_modeling_t5.py:1]: Please remove this file, it is irrelevant for mt5

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 5]
```diff
@@ -0,0 +1,662 @@
...
+from torch.nn import CrossEntropyLoss
+from transformers.modeling_outputs import (
+    BaseModelOutput,
+    BaseModelOutputWithPastAndCrossAttentions,
+    Seq2SeqLMOutput,
+)
+from transformers.models.mt5.modeling_mt5 import __HEAD_MASK_WARNING_MSG
+from transformers.utils import (
+    logging,
+)
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:32]:
```suggestion
from transformers.utils import logging
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 6]
```diff
@@ -0,0 +1,662 @@
...
+
+
+logger = logging.get_logger(__name__)
+
+try:
+    from habana_frameworks.torch.hpex.normalization import FusedRMSNorm as FusedRMSNorm
+except ImportError:
+    print("Not using HPU fused kernel for RMSNorm")
+    FusedRMSNorm = None
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:41]:
```suggestion
try:
    from habana_frameworks.torch.hpex.normalization import FusedRMSNorm as FusedRMSNorm
    
    has_fused_rms_norm = True
except ImportError:
    has_fused_rms_norm = False
    print("Not using HPU fused kernel for RMSNorm")
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 7]
```diff
@@ -0,0 +1,662 @@
...
+    """
+    Copied from https://github.com/huggingface/transformers/blob/v4.40.2/src/transformers/models/mt5/modeling_mt5.py 
+    The only differences are:
+        - override RMSNorm with Habana fused RMSNorm
+    """
+    # MT5 uses a layer_norm which only scales and doesn't shift, which is also known as Root Mean
+    # Square Layer Normalization https://arxiv.org/abs/1910.07467 thus varience is calculated
+    # w/o mean and there is no bias. Additionally we want to make sure that the accumulation for
+    # half-precision inputs is done in fp32
+    if hidden_states.device.type == "hpu" and FusedRMSNorm:
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:53]:
```suggestion
    if hidden_states.device.type == "hpu" and has_fused_rms_norm:
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 8]
```diff
@@ -0,0 +1,662 @@
...
+        layer_head_mask=None,
+        query_length=None,
+        use_cache=False,
+        output_attentions=False,
+        token_idx=None,
+    ):
+        """
+        Copied from https://github.com/huggingface/transformers/blob/v4.40.2/src/transformers/models/mt5/modeling_mt5.py
+        Self-attention (if key_value_states is None) or attention over source sentence (provided by key_value_states).
+        """
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:83]:
```suggestion
        The only differences are:
        - add new args token_idx
        """
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 9]
```diff
@@ -0,0 +1,662 @@
...
+        key_value_states=None,
+        position_bias=None,
+        past_key_value=None,
+        layer_head_mask=None,
+        query_length=None,
+        use_cache=False,
+        output_attentions=False,
+        token_idx=None,
+    ):
+        """
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:80]:
```suggestion
):
    """
```

There is an extra indentation in the whole function. Is there a reason for that, or just your api caused this. Please fix this extra indentation

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+                    hidden_states = past_key_value.index_copy_(-2, token_idx - 1, hidden_states)
+              elif past_key_value.shape[2] != key_value_states.shape[1]:
+                # checking that the `sequence_length` of the `past_key_value` is the same as
+                # the provided `key_value_states` to support prefix tuning
+                # cross-attn
+                # (batch_size, n_heads, seq_length, dim_per_head)
+                hidden_states = shape(proj_layer(key_value_states))
+              else:
+                # cross-attn
+                hidden_states = past_key_value
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:138]:
```suggestion
            if past_key_value is not None:
                if key_value_states is None:
                    # self-attn
                    # (batch_size, n_heads, key_length, dim_per_head)
                    if token_idx is None:
                        hidden_states = torch.cat([past_key_value, hidden_states], dim=2)
                    else:
                        hidden_states = past_key_value.index_copy_(-2, token_idx - 1, hidden_states)
                elif past_key_value.shape[2] != key_value_states.shape[1]:
                    # checking that the `sequence_length` of the `past_key_value` is the same as
                    # the provided `key_value_states` to support prefix tuning
                    # cross-attn
                    # (batch_size, n_heads, seq_length, dim_per_head)
                    hidden_states = shape(proj_layer(key_value_states))
                else:
                    # cross-attn
                    hidden_states = past_key_value
```

Please ensure consistent indentation throughout the entire code.

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+
+        if past_key_value is not None:
+            if len(past_key_value) != 2:
+                raise ValueError(
+                    f"past_key_value should have 2 past states: keys and values. Got { len(past_key_value)} past states"
+            )
+            if token_idx is None:
+              real_seq_length += past_key_value[0].shape[2] if query_length is None else query_length
+            else:
+              real_seq_length = past_key_value[0].shape[2]
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:99]:
```suggestion
            if token_idx is None:
                real_seq_length += past_key_value[0].shape[2] if query_length is None else query_length
            else:
                real_seq_length = past_key_value[0].shape[2]
```

Please ensure consistent indentation throughout the entire code.

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+            else:
+                position_bias = self.compute_bias(real_seq_length, key_length, device=scores.device)
+
+            # if key and values are already calculated
+            # we want only the last query position bias
+            if past_key_value is not None:
+             if token_idx is None:
+                position_bias = position_bias[:, :, -hidden_states.size(1) :, :]
+             else:
+                position_bias = position_bias.index_select(-2, token_idx - 1)
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:173]:
```suggestion
            if past_key_value is not None:
                if token_idx is None:
                    position_bias = position_bias[:, :, -hidden_states.size(1) :, :]
                else:
                    position_bias = position_bias.index_select(-2, token_idx - 1)
```

Please ensure consistent indentation throughout the entire code.

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 13]
```diff
@@ -0,0 +1,662 @@
...
+    layer_head_mask=None,
+    past_key_value=None,
+    use_cache=False,
+    output_attentions=False,
+    token_idx=None,
+):
+    """
+    Copied from https://github.com/huggingface/transformers/blob/v4.40.2/src/transformers/models/mt5/modeling_mt5.py
+    Self-attention (if key_value_states is None) or attention over source sentence (provided by key_value_states).
+    """
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:225]:
```suggestion
    The only differences are:
    - add new args token_idx
    """
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+        encoder_attention_mask=None,
+        encoder_decoder_position_bias=None,
+        layer_head_mask=None,
+        cross_attn_layer_head_mask=None,
+        past_key_value=None,
+        use_cache=False,
+        output_attentions=False,
+        return_dict=True,
+        token_idx=None,
+    ):
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:257]:
```suggestion
    ):
    """
    The only differences are:
    - add new args token_idx
    """
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 15]
```diff
@@ -0,0 +1,662 @@
...
+        if past_key_value is not None:
+            if not self.is_decoder:
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:259]:
```suggestion
    if past_key_value is not None:
        if not self.is_decoder:
```
Please ensure consistent indentation throughout the entire code.

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 16]
```diff
@@ -0,0 +1,662 @@
...
+        cross_attn_head_mask=None,
+        past_key_values=None,
+        use_cache=None,
+        output_attentions=None,
+        output_hidden_states=None,
+        return_dict=None,
+        token_idx=None,
+    ):
+
+        use_cache = use_cache if use_cache is not None else self.config.use_cache
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:346]:
```suggestion
    ):
    """
    The only differences are:
    - add new args token_idx
    """
    use_cache = use_cache if use_cache is not None else self.config.use_cache
```

There is an extra indentation in this function!
Please ensure consistent indentation throughout the entire code.

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 17]
```diff
@@ -0,0 +1,662 @@
...
+
+        batch_size, seq_length = input_shape
+
+        # required mask seq length can be calculated via length of past
+        if token_idx is not None:
+            mask_seq_length = past_key_values[0][0].shape[2] if past_key_values is not None else seq_length
+        else:
+            mask_seq_length = past_key_values[0][0].shape[2] + seq_length if past_key_values is not None else seq_length
+
+        if use_cache is True:
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:380]:
```suggestion
        if use_cache:
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 18]
```diff
@@ -0,0 +1,662 @@
...
+                    cross_attn_layer_head_mask=cross_attn_layer_head_mask,
+                    past_key_value=past_key_value,
+                    use_cache=use_cache,
+                    output_attentions=output_attentions,
+                    token_idx=token_idx,
+                )
+
+            # layer_outputs is a tuple with:
+            # hidden-states, key-value-states, (self-attention position bias), (self-attention weights), (cross-attention position bias), (cross-attention weights)
+            if use_cache is False:
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:469]:
```suggestion
            if not use_cache:
```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 19]
```diff
@@ -0,0 +1,662 @@
...
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        decoder_inputs_embeds: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        token_idx: Optional[torch.LongTensor] = None,
+    ) -> Union[Tuple[torch.FloatTensor], Seq2SeqLMOutput]:
+
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:536]:
```suggestion
:
    """
    The only differences are:
    - add new args token_idx
    """

```

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 20]
```diff
@@ -0,0 +1,662 @@
...
+    attention_mask=None,
+    head_mask=None,
+    decoder_head_mask=None,
+    decoder_attention_mask=None,
+    cross_attn_head_mask=None,
+    use_cache=None,
+    encoder_outputs=None,
+    token_idx=None,
+    **kwargs,
+):
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:634]:
```suggestion
    ):
    """
    The only differences are:
    - add new args token_idx
    """
```

**Comment by yafshar** [comment]
> @yafshar
> 
> python run_translation.py --model_name_or_path mt5-small --do_eval --source_lang en --target_lang ro --source_prefix "translate English to Romanian: " --dataset_name wmt16 --dataset_config_name ro-en --output_dir /tmp/tst-translation --per_device_eval_batch_size 4 --overwrite_output_dir --predict_with_generate --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --gaudi_config_name Habana/t5 --ignore_pad_token_for_loss False --pad_to_max_length --bf16
> 
> Is this you are asking? Please correct me.

@Gaurav7888 do you have the GPU performance number on A100 or others?

I was able to run with minor modification `--model_name_or_path google/mt5-small`

**Comment by yafshar** [comment]
This is what I currently get
```python
>> python run_translation.py --model_name_or_path google/mt5-small --do_eval --source_lang en --target_lang ro --source_prefix "translate English to Romanian: " --dataset_name wmt16 --dataset_config_name ro-en --output_dir /root/tst-translation --per_device_eval_batch_size 4 --overwrite_output_dir --predict_with_generate --use_habana --use_lazy_mode --use_hpu_graphs_for_inference --gaudi_config_name Habana/t5 --ignore_pad_token_for_loss False --pad_to_max_length --bf16

***** eval metrics *****
  eval_bleu                   =     0.0034
  eval_gen_len                =      127.0
  eval_loss                   =    51.5938
  eval_runtime                = 0:01:22.44
  eval_samples                =       1999
  eval_samples_per_second     =     24.247
  eval_steps_per_second       =      6.065
  max_memory_allocated (GB)   =       2.65
  memory_allocated (GB)       =       2.65
  total_memory_available (GB) =      94.62
```

File: optimum/habana/transformers/models/t5/modeling_t5.py
Comment (yafshar) [inline optimum/habana/transformers/models/t5/modeling_t5.py:1]: Please revert this commit. Irrelevant for mt5

**Comment by yafshar** [comment]
@Gaurav7888 can you address the comment to finish this PR?

File: optimum/habana/transformers/models/t5/modeling_t5.py [hunk 21]
```diff
@@ -507,6 +507,7 @@ def gaudi_T5ForConditionalGeneration_forward(
     return_dict: Optional[bool] = None,
     token_idx: Optional[torch.LongTensor] = None,
 ) -> Union[Tuple[torch.FloatTensor], Seq2SeqLMOutput]:
+
```
Comment (Gaurav7888) [inline optimum/habana/transformers/models/t5/modeling_t5.py:510]:
```suggestion
```

**Comment by Gaurav7888** [comment]
> @Gaurav7888 can you address the comments to finish this PR?

I have done the changes, please check

**Comment by yafshar** [comment]
@Gaurav7888 I will check the latest changes

File: optimum/habana/transformers/models/mt5/modeling_mt5.py [hunk 22]
```diff
@@ -0,0 +1,680 @@
...
+    Seq2SeqLMOutput,
+)
+from transformers.models.mt5.modeling_mt5 import __HEAD_MASK_WARNING_MSG
+from transformers.utils import logging
+
+
+logger = logging.get_logger(__name__)
+
+try:
+    from habana_frameworks.torch.hpex.normalization import FusedRMSNorm as FusedRMSNorm
```
Comment (yafshar) [inline optimum/habana/transformers/models/mt5/modeling_mt5.py:36]:
```suggestion
    from habana_frameworks.torch.hpex.normalization import FusedRMSNorm

```

**Comment by yafshar** [comment]
@Gaurav7888 there are `IndentationError` introduced by unexpected indentation in the code. Please check and fix them

**Comment by yafshar** [comment]
@Gaurav7888 would you please address the comment to wrap up this PR

**Comment by yafshar** [comment]
@Gaurav7888 any update?

**Comment by libinta** [comment]
@Gaurav7888  I am closing the PR as no response

**Pull request closed without merging**